### ISSUE TITLE
Upgrading from v3 to v4

### ISSUE THREAD

**Issue by philtgarner** [graphaware/neo4j-php-client#26]
I'm currently trying to update from v3 (neoxygen/neo4j-neoclient) to v4 and I'm having quite a few issues, potentially due to the apparent lack of documentation for the new version (perhaps I'm looking in the wrong place?).

In particular, I'm cannot work out how to handle queries that return nodes rather than tabular information. i.e. `MATCH (p:Person) RETURN p` rather than `MATCH (p:Person) RETURN p.name, p.age`. In v3 I made use of methods like `$person = $results->getSingleNode('Person')` and `$person->getConnectedNodes('IN', 'FRIENDS')`. I cannot seem to see how this is possible in the new version.

I'm sure the improvements of going from v3 to v4 are worth while but the transition process doesn't seem that clear to me.

**Comment by ikwattro** [comment]
Hi @philtgarner 

Thanks for the feedback, my reply here will then serve later as addition to the README for people upgrading.

We will talk then about `Handling Results` : 

There are 3 main concepts around this topic : 

1. a Result
2. a Record
3. a RecordValue

Let's take a look at a query we do in the browser containing multiple possibilities of types : 

```
MATCH (n:Address)
RETURN n.address as addr, n, collect(id(n)) as ids
LIMIT 5
```

![screen shot 2016-05-11 at 20 54 34bis](https://cloud.githubusercontent.com/assets/1222009/15192806/1a39cb30-17bb-11e6-8687-ed861411af2d.png)


#### Result

A `Result` is a collection of `Record` objects, every **row** you see in the browser is a `Record` and contains `Record Value`s.

* In blue the Result
* In orange a Record
* In green a RecordValue

#### Record

In contrary to the previous versions of the client, there is no more automatic merging of all the records into one big record, so you will need to iterate all the records from the `Result` : 

```php
$query = "MATCH (n:Address)
RETURN n.address as addr, n, collect(id(n)) as ids
LIMIT 5";
$result = $client->run($query);

foreach ($result->records() as $record) {
  // here we do what we want with one record (one row in the browser result)
  print_r($record);
}
```

#### Record Value

Every record contains a collection of `Record Value`s, which are identified by a `key`, the key is the identifier you give in the `RETURN` clause of the Cypher query. In our example, a `Record` will contain the following keys : 

* addr
* n
* ids

In order to access the value, you make use of the `get()` method on the `Record` object : 

```php
$address = $record->get('addr');
```

The type of the value is depending of what you return from Neo4j, in our case the following values will be returned :

* a `string` for the `addr` value
* a `Node` for the `n` value
* an `array` of `integers` for the `ids` value

Meaning that :

```php
$record->get('addr'); // returns a string
$record->get('n'); // returns a Node object
$record->get('ids'); // returns an array
```

`Node`, `Relationship` and `Path` objects have then further methods, so if you know that the node returned by the identifier `n` has a `countries` property on it, you can access it like this :

```php
$addressNode = $record->get('n');
$countries = $addressNode->value('countries');
```

The `Record` object contains three methods for IDE friendlyness, namely :

```php
$record->nodeValue('n');
$record->relationshipValue('r');
$record->pathValue('p');
```

This does not offer something extra, just that the docblocks hint the IDE for autocompletion.


#### Extra: ResultCollection

When you use `Stack` objects for sending multiple statements at once, it will return you a `ResultCollection` object containing a collection of `Result`s. So you need to iterate the results before accessing the records.

--- 

Let me know if there is something missing here or something that can still be improved.

Cheers !

**Comment by philtgarner** [comment]
Thanks for such a detailed reply, it really has helped.

I think it'll benefit the wider community to include this in the readme for future reference. I think, to summarise, when working with this new version you should visualise the results of a query using "Rows" rather than "Graph" in the Neo4j browser.

One small question, what is the difference between `$client->run($query)->records()` and `$client->run($query)->getRecords()`. The readme refers to `getRecords()` and you mentioned `records()` in your reply. Is there a difference and which would you recommend?

Many thanks

**Comment by ikwattro** [comment]
1) Yes it will definitely go as a `from v3` to `v4` in the readme. 

2) Yes, it should be better see as rows than graphs, a not on this is that with the release of Neo4j 3.0, I followed the official Neo4j Technology Compliance Kit for Unified Drivers (still private), the benefit is that code examples you will see in the future will be unified across driver languages (except some language specific differences).

This doesn't mean however I cannot add functionality to the client, for example `$result->toGraph()` in the future.

There is no difference, the former is the official method name from the TCK, the latter is the first method I used before the TCK was including the Result API. I didn't removed it but will definitely do it in v5 (don't worry, it's not for now :) )

In overall, I think that by keeping a compliance with the TCK, migrations will be easier in the future. And for us as Neo4j partners and consultants it makes sense for enterprise support.
